syntax = "proto3";

option cc_enable_arenas = true;

package allspark;


message AuthData{
   string node_id = 1;
   
   int32 client_type  = 2; // 0 = normal router mode exchanging protobuf data |  1 = raw publisher mode | 2 = raw subscriber mode
}

message Message {
     string dest_id = 1;
     string src_id = 2;
 	 string dest_session_id = 3;
     int32 type =  4 ;
     bytes payload = 5;
     bytes payload1 = 6 ;
     bytes payload2 = 7 ;
     string id = 8 ;
     bool is_ack_required = 9 ;
     bool is_anonymize_src_id = 10;
     string anonymize_src_id = 11;
     int64  timestamp = 12;
     bytes raw_data = 13;
     
}

message InboxMessage{
	string node_id_seq =1 ;
	string from_id =2 ;
	int32 message_type = 3;
	string payload = 4;
	int64 timestamp = 5;
}

message ConfigRequest{
	string update_gcm_key = 1;
	bool fetch_inbox_messages = 2;
	int64 timestamp = 3;
	int64 fetch_inbox_messages_from_seq = 4;
	int64 fetch_inbox_messages_to_seq = 5;
	int64 fetch_inbox_messages_from_timestamp = 6;
	int64 fetch_inbox_messages_to_timestamp = 7;
}

message ConfigResponse{
	repeated InboxMessage messages = 1;
	int64 from_seq = 2;
	int64 to_seq = 3;
	bool more =4;
	int64 user_time_stamp_diff = 5;
}



message NodeData{
	string host_address = 1;
	string port = 2;
	bool is_ssl = 3;
}

message SessionNodeData{
	string node_id = 1;
	bool is_anonymous = 2;	
}

message SessionData{
	string session_id = 1;
	string description = 2;
	string node_id = 3;
	repeated SessionNodeData session_nodes = 4;
	
	bool allow_anonymous = 5;
}

message SessionsList{
	repeated SessionData sessions = 1;
	int32 from = 2;
	int32 count = 3;
}

message NodesListData{
	repeated string node_id = 1;
}



//requests and responses

message CreateSessionRequest{
	
	
	bool is_anonymous = 1;	
	
	int32 who_can_add_session_nodes = 2;
	int32 notify_only_last_few_users = 3;
	
	int32 session_type = 4;
	string session_game_master_node_id = 5;
	
	string session_description = 6;
	bool allow_anonymous = 7;
}

message JoinSessionRequest{
	bool is_anonymous = 1;	
	string session_id = 2;
	string node_id = 3;
}





message GetInboxMessagesRequest{
	string session_id = 1;	
	string node_id = 6;
	
	int32 from_seq = 2;
	int32 to_seq = 3;
	int64 from_timestamp = 4;
	int64 to_timestamp = 5;
}

message InboxMessagesResponse{
	repeated InboxMessage inbox_messages = 1;
	int64 from_seq = 2;
	int64 to_seq = 3;
	bool more =4;
}

message BooleanData{
	bool value  = 1;
}

message StringData{
	string value  = 1;
}

message EmptyData{
}

/*
../protobuf/src/protoc --cpp_out=./wire_messages/ ./wire_transfers.proto && \
../protobuf/src/protoc --python_out=./python_utils/ ./wire_transfers.proto && \
../protobuf/src/protoc --java_out=./java_utils/ ./wire_transfers.proto
*/
